src/main/java/com/google/sps/servlets/TripsServlet.java

```diff
@@ -0,0 +1,197 @@
+      Map<Long, List<TripLocation>> tripLocationMap = new HashMap<>();
+      for(Entity entity : tripLocationEntityIterable) {
+        TripLocation location = convertEntityToTripLocation(entity);
+        if(tripLocationMap.containsKey(location.timestamp())) {
+          List<TripLocation> locationsUnderCurrTrip = tripLocationMap.get(location.timestamp());
+          locationsUnderCurrTrip.add(location);
+          tripLocationMap.replace(location.timestamp(), locationsUnderCurrTrip);
+        } else {
+          tripLocationMap.put(location.timestamp(), new ArrayList<>(Arrays.asList(location)));
+        }
+      }
+
+      // Get list of trips without their corresponding locations
+      Map<Trip, List<TripLocation>> userTripsDataResponse = new HashMap<>();
+      for (Entity entity : tripEntityIterable) {
+        Trip trip = convertEntityToTrip(entity);
+        userTripsDataResponse.put(trip, tripLocationMap.get(trip.timestamp()));
+      }
+
+      response.setContentType("application/json;");
+      Gson gson = new Gson();
+      String serializedJSON = gson.toJson(userTripsDataResponse);
+      response.getWriter().println(serializedJSON);
+      response.setStatus(HttpServletResponse.SC_OK);
+    } else {
+      // send error code 401 if user is not authenticated and tries to access data
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json;");
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String userEmail = userService.getCurrentUser().getEmail();
+      long timestamp = System.currentTimeMillis();
+      
+      Trip trip = Trip.builder()
+                    .setTitle(request.getParameter(Trip.ENTITY_PROPERTY_TITLE))
+                    .setHotel(request.getParameter(Trip.ENTITY_PROPERTY_HOTEL))
+                    .setRating(Double.valueOf(request.getParameter(Trip.ENTITY_PROPERTY_RATING)))
+                    .setDescription(request.getParameter(Trip.ENTITY_PROPERTY_DESCRIPTION))
+                    .setOwner(userEmail)
+                    .setTimestamp(timestamp)
+                    .build();
+      Entity tripEntity = convertTripToEntity(trip);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(tripEntity);
+
+      // build TripLocation objects from request data and put them in Datastore
+      
+      String[] locations = request.getParameterValues("locations");
+      String[] weights = request.getParameterValues("weights");
+      // these two arrays should be of equal length. If not, return error
+      if(locations.length != weights.length) {
+        response.setStatus(HttpServletResponse.SC_CONFLICT);
+        return;
+      }
+      
+      for(int i = 0; i < locations.length; i++) {
```

Comment:
Checkout the "parent" key in the Entity: https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Entity.html#Entity-java.lang.String-java.lang.String-com.google.appengine.api.datastore.Key-

This might gives you an option to store the TripLocations under a specific Trip, and you can query the triplocation under a certain parent.

NoSQL database usually supports such structure.

Reply:
Done. This is definitely a less hacky solution, thanks for pointing it out.